fix: use SAL2 `_In_`, `_Out_` to prevent conflicts

ShortDevelopment opened the pull request:
CC uses the [Microsoft source-code annotation language (SAL)](https://learn.microsoft.com/en-us/cpp/code-quality/understanding-sal) to enable static analysis of pointer usage etc.

The CC code-base still uses the old annotations (e.g. `__in` instead of `_In_`).
Some of the old SAL annotation (i.e. `__in`) stand in conflict with the stdlib.

Here's an inofficial guide: https://samscode.blogspot.com/2015/02/sal-to-sal2-porting-guide.html

### Changes
- Replace `\b__in\b` with `_In_` and `\b__out\b` with `_Out_`
  https://github.com/dotnet/runtime/pull/63261
- Use Ubuntu 24.04 instead of 20.04 in ci
  (Should this be a separate PR?)
- Ignore `-Wnontrivial-memcall` to allow `memset` with non-`void*`

---

@rhuanjl

ShortDevelopment committed:
Remove old SAL definitions

ShortDevelopment committed:
`\b__in\b` ⇾ `_In_ `

ShortDevelopment committed:
`\b__out\b` ⇾ `_Out_ `

ShortDevelopment committed:
Fix copyright

ShortDevelopment committed:
Fix eol

rhuanjl commented:
Good to see this clean up.

I don't like all the warning removals the project already has rather than increasing them can we change the code?
Is the memset with non-void* essential? Could we get rid of the warning by casting the relevant pointer(s)?

ShortDevelopment commented:
@rhuanjl Sure. The compiler now enforces an explicit `void*` cast for `memset` etc.
I can fix that later today.

ShortDevelopment committed:
Explicit `(void*)` for non-trivial `memset`

first argument in call to 'memset' is a pointer to non-trivially copyable type [-Wnontrivial-memcall]

ShortDevelopment committed:
Fix copyright

rhuanjl commented on azure-pipelines.yml:
```diff
@@ -52,13 +52,13 @@ jobs:
           libtype_flag: ''
-        Ubuntu20.ReleaseWithDebug:
-          image_name: 'ubuntu-20.04'
+        Ubuntu24.ReleaseWithDebug:
```
The 4 Linux tests above were meant to be latest ubuntu then these two below were meant to be an older version for slightly broader coverage.

To keep the pattern should put the top 4 to Ubuntu24 and change these two to Ubuntu 22.

ShortDevelopment replied:
Good point!
Done in 88fc0611f3e432709f74555caf16cb48f0043239 (I removed the old commit through rebase)

rhuanjl commented:
Happy with everything other than the CI change (see comment above).

ShortDevelopment committed:
Ubuntu 24
